end2end-test-examples/echo-client/src/main/java/io/grpc/echo/EchoClient.java

```diff
@@ -51,10 +63,43 @@ public class EchoClient {
   private Map<String, Map<Boolean, AtomicLong>> errorCounts = new ConcurrentHashMap<>();
   private final String OTHER_STATUS = "OTHER";
 
+  private static final Tracer tracer = Tracing.getTracer();
+
+  private final ClientInterceptor peerCheckInterceptor =
+      new ClientInterceptor() {
+        @Override
+        public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+            MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+          final ClientCall<ReqT, RespT> clientCall = next.newCall(method, callOptions);
+          return new SimpleForwardingClientCall<ReqT, RespT>(clientCall) {
+            @Override
+            public void start(Listener<RespT> responseListener, Metadata headers) {
+              super.start(
+                  new SimpleForwardingClientCallListener<RespT>(responseListener) {
+                    @Override
+                    public void onHeaders(Metadata headers) {
+                      // Check peer IP after connection is established.
+                      SocketAddress remoteAddr =
+                          clientCall.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+                      String peerIP =
+                          ((InetSocketAddress) remoteAddr).getAddress().getHostAddress();
+                      System.out.println("Peer IP = " + peerIP);
```

Comment:
Are we just printing peer address for debugging?

Reply:
Yeah that is the way to tell if DirectPath is used. This peer IP only applies when DirectPath is used.